components/core/src/clp_s/SingleFileArchiveDefs.hpp

```diff
@@ -0,0 +1,56 @@
+#ifndef CLP_S_ARCHIVEDEFS_HPP
+#define CLP_S_ARCHIVEDEFS_HPP
+
+#include <string>
+
+#include "msgpack.hpp"
+
+namespace clp_s {
+// define the version
+constexpr uint8_t cArchiveMajorVersion = 0;
+constexpr uint8_t cArchiveMinorVersion = 2;
+constexpr uint16_t cArchivePatchVersion = 0;
+
+struct ArchiveHeader {
+    uint8_t magic_number[8];
+    uint32_t version;
+    uint64_t uncompressed_size;
+    uint64_t compressed_size;
+    uint64_t reserved_padding[4];
+    uint32_t metadata_section_size;
+    uint16_t compression_type;
+    uint16_t padding;
+};
+
+enum class ArchiveCompressionType : uint16_t {
+    Zstd = 0,
+};
+
+enum struct ArchiveMetadataPacketType : uint8_t {
+    ArchiveInfo = 0,
+    ArchiveFileInfo = 1,
+    TimestampDictionary = 2,
+};
+
+struct ArchiveInfoPacket {
+    uint64_t num_segments;
+    // TODO: Add more fields in the future
+
+    MSGPACK_DEFINE(num_segments);
+};
+
+struct ArchiveFileInfo {
+    std::string n;  // name
+    uint64_t o;  // offset
+
+    MSGPACK_DEFINE(n, o);
+};
+
+struct ArchiveFileInfoPacket {
+    std::vector<ArchiveFileInfo> files;
+
+    MSGPACK_DEFINE(files);
```

Comment:
We probably want to switch these to MSGPACK_DEFINE_MAP. Default behaviour of MSGPACK_DEFINE is to serialize everything as nested msgpack arrays of just the values without including the keys. As well, I'm unable to deserialize the msgpack bytes that this current approach generates with the python msgpack library (though I did confirm that the encoded bytes were valid msgpack via manual inspection). I'm also unable to use the unpacking code generated by this macro to deserialize the generated bytes at the moment.

Reply:
Actually upon further inspection, if I remove the first two bytes of the stream I seem to be able to deserialize succesfully. (First two bytes specify a msgpack array of length 1 containing the integer '1' for some reason).

Reply:
Figured it out. There's a bug in the metadata writing code. Will add a comment there.

Reply:
Even after fixing that bug locally the deserializer code provided by MSGPACK_DEFINE does not for ArchiveFileInfoPacket. I'm able to hack around it in a fragile way, but we should check if MSGPACK_DEFINE is allowed to be used in a nested way like what we do here, and also check whether std::vector is fully supported.